devices/src/virtio/queue.rs

```diff
@@ -375,6 +413,19 @@ impl Queue {
     pub fn go_to_previous_position(&mut self) {
         self.next_avail -= Wrapping(1);
     }
+
+    /// Fetch the available ring index (`virtq_avail->idx`) from guest memory.
+    /// This is written by the driver, to indicate the next slot that will be filled in the avail
+    /// ring.
+    fn avail_idx(&self, mem: &GuestMemory) -> Wrapping<u16> {
+        // Bound checks for queue inner data have already been performed, at device activation time,
+        // via `self.is_valid()`, so it's safe to unwrap and use unchecked offsets here.
+        // Note: the `MmioDevice` code ensures that queue addresses cannot be changed by the guest
+        //       after device activation, so we can be certain that no change has occured since
+        //       the last `self.is_valid()` check.
+        let addr = self.avail_ring.unchecked_add(2);
```

Comment:
Not sure where the `2` is from.

Reply:
This is `offsetof(struct virtq_avail, idx)`. See my previous comment regarding `struct virtq_avail` and offset arithmetic.